### ISSUE TITLE
Filter empty masks.

### ISSUE THREAD

**Issue by maxfrei750** [albumentations-team/albumentations#745]
## 🐛 Bug

Maybe it is not a bug but a feature or I just didn't find the right keyword to achieve the behavior that I would expect.

If there is a sample with multiple annotations (e.g. multiple bboxes and masks) and a part of the image is removed by `RandomCrop`, then bboxes outside of the cropped image and the corresponding labels are removed (to be expected). However, the corresponding masks remain as now-empty masks.

## To Reproduce

1. Take a sample with multiple annotations, consisting of bboxes, labels and masks.
1. Crop it using `RandomCrop`, so that some instances are no longer inside the image.
1. Check the number of masks and bboxes. They are not the same.

```
transform = A.Compose([
    A.RandomCrop(width=100, height=100),
    ], bbox_params=A.BboxParams(label_fields=["class_labels"]))

data = self.transforms(image=image, masks=masks, bboxes=boxes, class_labels=labels)

assert(len(data["masks"]) == len(data["bboxes"], "Different numbers of bboxes and masks")
```
## Expected behavior

I think that empty masks should be removed.

**Comment by Dipet** [comment]
It is very strange behavior to remove empty mask, because when you train segmentation model you need mask. In common if mask empty all values will be equal to 0.

**Comment by maxfrei750** [comment]
Good point. However, I'm dealing with instance segmentation, where you have multiple masks per image.

**Comment by BloodAxe** [comment]
For instance segmentation you can use one image of uint16 dtype, where each instance object is encoded with unique value. Of course, this would work only for non-overlapping instances.

**Comment by maxfrei750** [comment]
> Of course, this would work only for non-overlapping instances.

That's exactly the problem.

**Comment by guimaraesmarcelo** [comment]
Hi @BloodAxe, I faced the same problem while training an instance segmentation model.  Would any solution be available yet?

In my case, I noticed that when I get empty masks, the number of empty masks is higher than the number of removed bounding boxes labels.

**Comment by maxfrei750** [comment]
@guimaraesmarcelo I did the following:

Assume that `transform` is your albumentations transform, `image` is your image, and `masks` are your masks.

```
transformed_data = transform(image=image, masks=masks)

image = transformed_data["image"]
masks = transformed_data["masks"]

 is_not_empty = [np.any(mask) for mask in masks]

masks = list(compress(masks, is_not_empty))
labels = list(compress(labels, is_not_empty))
scores = list(compress(scores, is_not_empty))

boxes = [extract_bounding_box(mask) for mask in masks]
```

With
```
def extract_bounding_box(mask) -> np.ndarray:
    """Extract the bounding box of a mask.
    :param mask: HxW numpy array
    :return: bounding box
    """
    pos = np.where(mask)

    if not (pos[0].size or pos[1].size):
        return np.array([0, 0, 0, 0])

    xmin = np.min(pos[1])
    xmax = np.max(pos[1]) + 1
    ymin = np.min(pos[0])
    ymax = np.max(pos[0]) + 1
    return np.array([xmin, ymin, xmax, ymax])
```

I hope that helps. For the complete code, please refer to: https://github.com/maxfrei750/paddle/blob/0b36b9e1543614f1800da297ccc54cf10b14f927/paddle/data/sample.py#L49

**Comment by guimaraesmarcelo** [comment]
Thanks @maxfrei750, I think I got the idea. You are basically getting the data that was transformed and has a valid mask and then reconstructing the bounding boxes. Sounds like a good hotfix.

**Comment by maxfrei750** [comment]
@guimaraesmarcelo Exactly. I hope it helps. :+1:

**Comment by aurelien-m** [comment]
I'm facing the same issue when using Mask RCNN from torchvision (instance segmentation). It could be nice to have an option to remove empty masks in that particular use case

**Comment by gsoykan** [comment]
I've faced the same issue and have a more straightforward solution. The idea is to use indices with label fields for transformation. Here is what it looks like:

```

transformations = A.Compose([
...
    ],
        bbox_params=A.BboxParams(format='pascal_voc', label_fields=[..., 'indices'], min_area=16,
                                 min_visibility=0.1))

...

sample = self.transform(image=image,
                                bboxes=target['boxes'],
                                masks=instance_masks,
                                ...,
                                indices=np.array(list(range(len(target['boxes'])))))

transformed_masks = np.array(sample['masks'])[sample['indices']]
```
